### PR TITLE
Protocol Fees (Set to 20%)

### PR DESCRIPTION
**Pull request by kgrgpg**
This is the implementation for incorporating the fees that will be reserved for protocol usage. The commission collected can be used for incentivising different stakeholders. The current fees is set to 20 % and hard coded to take into account the format of decimals in solidity language. This code only aims to hold back the commission and does not show how to utilise or withdraw it from the contract at the moment. The issue for the same can be found at https://github.com/vacp2p/rfc/issues/485

### CODE REVIEW

**Review by richard-ramos — APPROVED** [review]

File: contracts/Rln.sol [hunk 1]
```diff
@@ -86,7 +88,8 @@ contract RLN {
 		members[_pubkeyIndex] = 0;
 
 		// refund deposit
-		(bool sent, _) = receiver.call{value: MEMBERSHIP_DEPOSIT}("");
+		uint256 withdrawableDeposit = MEMBERSHIP_DEPOSIT * (WAD - PROTOCOL_FEES_PERCENTAGE) / WAD
```
Comment (staheri14) [inline contracts/Rln.sol:91]: To dave on the gas, isn't it better to do this calculation once (e.g., in the constructor or when defining the member fields) and save the result in a variable like 'withdrawable_amount`  and then reuse the calculated value in this function call?

File: contracts/Rln.sol [hunk 2]
```diff
@@ -86,7 +88,8 @@ contract RLN {
...
+		(bool sent, _) = receiver.call{value: withdrawableDeposit}("");
```
Comment (staheri14) [inline contracts/Rln.sol:92]: Please also add comments next to each code line to elaborate on what is happening.

**Comment by staheri14** [comment]
Also, following our call, please create an issue for passing the fee percentage as a constructor argument (instead of being hardcoded).

File: contracts/Rln.sol (hunk 1 above)
Comment (staheri14) [reply contracts/Rln.sol:91]: Great! Thanks for the insights, the result is so unexpected! 
Considering this observation, do you think it is more cost-efficient to keep the `PROTOCOL_FEES_PERCENTAGE` as a hardcoded value or to pass it as an argument to the constructor in which case it becomes an immutable variable (hence it causes a higher execution cost when it comes to the `withdraw` function. 
I already spoke with @kgrgpg and he is going to inspect ENS contract to see how the service fee is modeled there.  NIn the meanwhile, would be also interested to know your inputs @richard-ramos

Comment (richard-ramos) [reply contracts/Rln.sol:91]: I'm not an expert in this topic, but for costs, I believe it should be more efficient to have the percentage hardcoded, but it has a disadvantage of not being able to change this value later unless whoever uses the contract also implements some mechanism like making the contract upgradeable which could allow the contract owner to change the implementation.

Comment (staheri14) [reply contracts/Rln.sol:91]: Just noticed that if the `PROTOCOL_FEES_PERCENTAGE` is set in the constructor, it does not have to be `mutable`, hence making it a parameter of the constructor would not cause any additional gas cost in the `withdraw` function.

File: contracts/Rln.sol (hunk 2 above)
Comment (staheri14) [reply contracts/Rln.sol:92]: > Please also add comments next to each code line to elaborate on what is happening.

@kgrgpg  Once the comment is addressed, let me know to re-review.

**Comment by kgrgpg** [comment]
@staheri14 The protocol fees setting via constructor issue can be found at https://github.com/vacp2p/rln-contract/issues/8

**Comment by rymnc** [comment]
Closing this PR as tokenomics hasn't been allotted to rlnp2p, yet

**Pull request closed without merging**